Fix break with LAYER_BASED_SVG_ENGINE disabled after 269492@main

ykimot opened the pull request:
#### b2f96217de848269bb771621e1c49a3323e65974
<pre>
Fix break with LAYER_BASED_SVG_ENGINE disabled after 269492@main
<a href="https://bugs.webkit.org/show_bug.cgi?id=263362">https://bugs.webkit.org/show_bug.cgi?id=263362</a>

Reviewed by Ross Kirsling.

When LAYER_BASED_SVG_ENGINE is disabled, RenderSVGShape is used in
SVGRenderSupport::calculateApproximateStrokeBoundingBox() and the class
is defined with LAYER_BASED_SVG_ENGINE enabled. Also an inline error
occurs in SVGRenderSupport.cpp.

To fix the build break, it needs to add a conditional code for
LAYER_BASED_SVG_ENGINE and add include a header file to refer RenderSVGShape.

* Source/WebCore/rendering/svg/SVGRenderSupport.cpp: Add include LegacyRenderSVGShapeInlines.h
(WebCore::SVGRenderSupport::calculateApproximateStrokeBoundingBox): Add a condtional macro for LAYER_BASED_SVG_ENGINE

Canonical link: <a href="https://commits.webkit.org/269511@main">https://commits.webkit.org/269511@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/38fd1c048d5a4717ba5fbcc0e099d008cee796b3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22797 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/613 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23884 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24705 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21102 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23054 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/1537 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23323 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22000 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23037 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/287 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19773 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25558 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/288 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20653 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/26859 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/20650 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/20902 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24715 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/340 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/18160 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/264 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5428 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/389 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/327 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rkirsling commented on Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
```diff
@@ -650,6 +652,10 @@ FloatRect SVGRenderSupport::calculateApproximateStrokeBoundingBox(const RenderEl
     ASSERT(is<RenderSVGShape>(renderer));
     const auto& shape = downcast<RenderSVGShape>(renderer);
     return shape.adjustStrokeBoundingBoxForZeroLengthLinecaps(RepaintRectCalculation::Fast, calculate(shape));
```
It seems that you should just guard these lines with `#if ENABLE(LAYER_BASED_SVG_ENGINE)` instead of recopying code into an `#else`—this would match the [earlier case on line 481](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/rendering/svg/SVGRenderSupport.cpp#L481-L484).

ykimot replied:
Right, I'd like to use the coding style but there is one unclear part there. 

[Source/WebCore/rendering/svg/SVGRenderSupport.cpp](https://github.com/WebKit/WebKit/commit/f7f897982b66445d96622a5b67145280e2ba2d81#diff-542b0605fe303b01bd02c8cd36c5e92cb3c5dee6ad232704ecb20bca9ea4f94fR645-R648
)
>     if (is<LegacyRenderSVGShape>(renderer)) {
>         const auto& shape = downcast<LegacyRenderSVGShape>(renderer);
>         return shape.adjustStrokeBoundingBoxForMarkersAndZeroLengthLinecaps(RepaintRectCalculation::Fast, calculate(shape));
>     }
> 

I think this part without "is&lt;LegacyRenderSVGShape&gt;(renderer)) would be a default case when LAYER_BASED_SVG_ENGINE is disabled. Following your suggestion, the part will be

```
#if ENABLE(LAYER_BASED_SVG_ENGINE)
    if (is<LegacyRenderSVGShape>(renderer)) {
#endif
        const auto& shape = downcast<LegacyRenderSVGShape>(renderer);
        return shape.adjustStrokeBoundingBoxForMarkersAndZeroLengthLinecaps(RepaintRectCalculation::Fast, calculate(shape));
#if ENABLE(LAYER_BASED_SVG_ENGINE)
    }
#endif

#if ENABLE(LAYER_BASED_SVG_ENGINE)
    ASSERT(is<RenderSVGShape>(renderer));
    const auto& shape = downcast<RenderSVGShape>(renderer);
    return shape.adjustStrokeBoundingBoxForZeroLengthLinecaps(RepaintRectCalculation::Fast, calculate(shape));
#endif
```

So it seems to make the part difficult to read, recopying the code is applied.  But I'm not sure whether my thinking is correct. If it's wrong, this part must be rewritten with a right design. I'd like to know how the part can be fixed through this review.

SVGRenderSupport::calculateApproximateStrokeBoundingBox() needs to return something when LAYER_BASED_SVG_ENGINE is disabled

rkirsling replied:
I was suggesting that you follow the earlier code exactly, which means letting the `if` stay, even though it's unnecessary. This policy makes the most sense since it is a build fix.

ykimot replied:
Thanks, I understand, will fix.

rkirsling replied:
Ah, sorry, I misunderstood. If all paths need to return a value then your original patch was best. Sorry for the confusion.

ykimot replied:
Don't mind. I could check this PR from wide variety of points. I appreciate your comments.

rkirsling reviewed (approved):
r=me

ykimot committed:
Fix break with LAYER_BASED_SVG_ENGINE disabled after 269492@main
https://bugs.webkit.org/show_bug.cgi?id=263362

Reviewed by Ross Kirsling.

When LAYER_BASED_SVG_ENGINE is disabled, RenderSVGShape is used in
SVGRenderSupport::calculateApproximateStrokeBoundingBox() and the class
is defined with LAYER_BASED_SVG_ENGINE enabled. Also an inline error
occurs in SVGRenderSupport.cpp.

To fix the build break, it needs to add a conditional code for
LAYER_BASED_SVG_ENGINE and add include a header file to refer RenderSVGShape.

* Source/WebCore/rendering/svg/SVGRenderSupport.cpp: Add include LegacyRenderSVGShapeInlines.h
(WebCore::SVGRenderSupport::calculateApproximateStrokeBoundingBox): Add a condtional macro for LAYER_BASED_SVG_ENGINE

Canonical link: https://commits.webkit.org/269511@main
